readme.txt
 Copyright (C) 2006 B.A.T.M.A.N. contributors:
 This program is free software; you can redistribute it and/or
 modify it under the terms of version 2 of the GNU General Public
 License as published by the Free Software Foundation.
 
 This program is distributed in the hope that it will be useful, but
 WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 General Public License for more details.
 
 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software
 Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 02110-1301, USA

 We would like to thank all people that donated their time and skills
 to this software!

Thomas Lopatic
Marek Lindner
Axel Neumann
Corinna 'Elektra' Aichele
Stefan Sperling
Felix Fietkau
Ludger Schmudde


*************************************************************************
The experimental version of batman. source files renamed to bmex.c.h
to avoid confusion.


Although the core routing concept is similar to the main batman
protocol this branch incorporates a number of experimental features
and a complete rewrite of the data structure maintained by each batman
node to keep track of received originator messages and identified
routes.

Also it offers a few additional information to a node (and its users) when trying to
find optimal metrics (and hazards) for selecting the best next hop towards the final
destination of a packet.

The current version aims to supports (this is very experimental or work in progress):

- Multi interface support
- Routing table showing selected and alternative routes and characteristics to each destination
- Accessible characteristics for each potential route include:
  - Next-hop ip
  - Timeout for bidirectional neighbor status of next hop
  - Interface
  - Number of total originator messages from destination received via this neighbor
  - Number of originator messages form destination received via this neighbor first
  - Average, minimm, and maximum TTL (Hops) of originator message
  - Minimum and maximum sequence number of received originator messages
  - Average, minimum, and maximum one-way trip time indicator which can be used to
    identify the fastest of possibly multiple routes to a destination.
- Originator packet bundling to reduce MAC overhead when (re-)broadcasting own and other
  nodes originator messages
- ...

*************************************************************************
some comments about the output:

showRoutes()  dev:ath0 ip:10.0.1.103 ttl:100 seq:167  dev:eth0.0 ip:10.0.5.103 ttl:1 seq:168 rmetric:1, cpolicy:3, fpolicy:2
to_orig         via_NB_orig      BDNB device all/1.  av/mi/ma-Ttl  mi/ma-SeqNo  av/mi/ma/la-ttpc
      10.0.1.76      10.0.5.102  2915 eth0.0  10  7  98.0  98  98    174   183     1.3    1    2    1
        alt. NB       10.0.1.76  2189   ath0   3  3  99.0  99  99    175   182     0.0    0    0    0
     10.0.1.102      10.0.5.102  2911 eth0.0  10  9  99.0  99  99    164   173     0.3    0    1    0
        alt. NB       10.0.1.76  2185   ath0   3  1  98.0  98  98    164   168     2.0    2    2    2
     10.0.5.102      10.0.5.102  2908 eth0.0  10 10   0.0   0   0    164   173     0.3    0    1    0





This batman (bmex) node has been started with two interface, namely
ath0 (ip 10.0.1.103) and eth0.0 (ip 10.0.5.103).  It is broadcasting
originator messages for 10.0.1.103 with a initial ttl of 100. It the
moment of the screenshot 167 Originator messages have been scheduled
for broadcasting on ath0 and eth0. Additionally this node is
broadcasting originator messages for 10.0.5.103 but only with a ttl of
one. This means the originator-device 10.0.5.103 is only viewable for
its direct neighbors. For this originator 168 originator messages have
been scheduled so fare. The node has been started with routing-metric
1, consider-policy 3 (specifies which originator messages should be
doroped or considered for further processing) and
forwarding/re-broadcasting-policy 2 (1-3-2 is currently the default
case but others could be implemented). 

The 1. line is followed by a list of potential routing entries to all
known nodes in the mesh. The first column shows the known destination
nodes. They are indicated by their ip address. If multiple routes to a
destination are known they are listed after the cuurent selected route
as alternative neighbor (alt. NB).

The 2. column shows the next-hop neighbor of the path to the
destination.

The 3. column shows the current timeout for the bidirectional link
status for this neighbor. -1 is shown if it is timed out.

The 4. column (BDNB) indicates the device in charge to reach that
neighbor.

The 5. column (all/1) shows the number of received
originator-sequence-number touples falling in the range of the 10
highest aware sequence numbers and the number of sequence numbers
first seen via this neighbor. As an example destination 10.0.1.76 is
reachable via two different neighbors (10.0.5.102 via eth0.0 and
directly via ath0). In this case the wireless link ath0 is less
reliable that the wired link (eth0.0) but it has one hop less than via
the wired link. Here all the last 10 sequencenumbers have been
received via eth0.0 but only 7 of them have been received via eth0.0
first. 3 of them have been received before and directly via link ath0.

  [ Note that the considered originator messages are not selected by a
  time frame but by the fact whether they belong to the highest 10
  sequence numbers aware of. Thus if a node successively received
  seqNo 1-20 (with 20 as the highes seqNo received so fare) it would
  show 10. If it received [1, 5, 11, 14, 20] it would show 3. If it
  successively received [1-14] and missed everything beyond it would
  also show 10. If no more originator messages are received after some
  time a garbage collector will come along and purge that destination
  from the list. This way route optimization can only be triggered by
  the reception of a new originator-sequencenumber touple and is never
  triggered by a timeout. I belive that unsynchronized timeouts on
  different nodes can be a reason for routing loops. ]

The 6. column (av/mi/ma-Ttl) shows the average, minimum, and maximum
ttl received with the last originator messages.

The 7. column (mi/ma-SeqNo) shows the minimum and maximum received
sequence numbers within the range of the last received originator
messages.

The 8. column (av/mi/ma/la-ttpc) aims to indicate the average,
minimum, maximum, and latest one-way trip time (plus an unknown
constant C per destination). Because the unknown constant C is equal
for all paths and originator messages received from a particular
destination, the indicated values can be used to estimate the fastest
path to a destination. 

  [ The acquirement of this value is a bit tricky. It is based on the
  idea that each node (re-)broadcasting its own and other node's
  originator messages keeps track of the INTENTIONAL delay (jitter)
  between reception and re-broadcasting of the message. Each node is
  doing so by adding this delay to a field (calles hold-bach-time) in
  the message itself. This way, every node receiving the same
  originator message from a distant node via two or more neighbors can
  combine the message's reception timestamp and hold-back-time field
  and use this information to evaluate which of the potential path to
  the destination suffers the least UNINTENTIONAL (processing, MAC, or
  PHY) delay. ]


CHANGELOG
For a more detailed and up-to-date CHANGELOG visit:
http://open-mesh.net/log/trunk/batman-experimental

Important changes should be listed in the following, most recent changes on top:


---------------------------------------------------------------------------------------------------
2009-08-18 - bmx_http_info and bmx_gsf_map plugin, ...

- simple http_info plugin. See lib/bmx_http_info/HOWTO

- decentral-openstreetmap-json-based map plugin from http://graciasensefils.net (GSF) folks

- add symbolic links from lib/bmx_*.so to lib/bmx_*/bmx_*.so.1 to 
  facilitate execution and debugging of local bmx libraries by 
  overwriting bmx plugin-library path. E.g.:
  BMX_LIB_PATH=$(pwd)/lib ./bmxd -d3

- Antenna-diversity support for bmxd seems working but unfortunately there are few wireless drivers which support 
  my understanding of the typical antenna-diversity implementation. This is what I hoped (maybe I am wrong): 
  - The RX-antenna is detected on-the-fly on a per-packet basis by comparing 
    the rcvd signal-strength via each antenna during reception of the phy-preamble.
  - The TX-antenna is determined per MAC-address based on the last detected best RX-antenna for this MAC.
  - Broadcast packets should be send round-robin like via each enabled TX-antenna (e.g. alternating via ant1 and ant2). 


---------------------------------------------------------------------------------------------------
2009-08-10 - bmx_uci_config plugin for openWrt universial configuraton interface (UCI)

- plugin for dynamic interaction with uci http://downloads.openwrt.org/sources/uci-0.7.3.tar.gz
- to compile first install uci: 
	wget http://downloads.openwrt.org/sources/uci-0.7.3.tar.gz
	tar xzvf uci-0.7.3.tar.gz; cd uci-0.7.3; make; sudo make install
- default configuration backend is: /etc/config/bmx
- see lib/bmx_uci_config/etc_config for a simple (bmx) and an advanced (bmx-advanced) example


---------------------------------------------------------------------------------------------------
2009-08-10 - refactoring scheduling, debugging, controlling,... introduce plugin framework, cleanup...

- new bmx help with -v / -h / -H / -x / -X

- new configuration features
	- generalize resetting of configured functions using "-", e.g.: -co- or -c dev -ath0
	--test to test remaining args without applying them
	--runtime-dir /var/run/bmx to store unix-socket and pid files
	--loop_mode / -l , make batch_mode default, -b deprecated, introduce -l for loop instead
	--throw ( was --no-tunnel-rule )
	--originators, --links, --hnas, --interfaces
	--purge_timeout 
	--dad_timeout
	--dev=ath0 and --dev=-ath0 (dynamically add and remove interfaces)
	-X show BatMan-eXperimental help, short description and configurable values
	--plugin bmx_plugin.so to load extended plugin functionality
	--plugin bmx_plugin.so -X or -cX to show available options including extended plugin functionality

- introduce section-like options
	--service=1.2.3.4:5 /seqno=1
	--dev=eth0 /ttl=4
	--unicast_hna=dhcp-clients /network=1.2.3.0 /netmask=24
	--throw=my-vpn /network=192.168.178.0 /netmask=24

- allow loopback interfaces as primary interfaces, e.g.: --dev lo:bmx

- new compile-time defines:
	-DNOTRAILER						- ~3K
	-DNODEBUGALL						- ~13k
	-DNOTUNNEL  	(only affects this node)		- ~23k
	-DNOSRV  	(only affects this node)		- ~3k
	-DNOVIS  	(only affects this node)		- ~2k
	-DNODEPRECATED	(for backward compatibility)		- ~2k

- new advanced compile-time defines (please dont touch):
	-DNOHNA		(affects all nodes in network)		- ~6k
	-DNOPARANOIA	(makes bug-hunting impossible)		- ~2k
	-DEXTDEBUG	(this eats your cpu)			+ ~0k
	-DTESTDEBUG	(testing syntax of __VA_ARGS__ dbg...() macros)
	-DWITHUNUSED	(includes yet unused stuff)


- allow full-broadcast destination addresses to facilitate IP renumbering

- track configured ip-rules and ip-routes and dont configure things twice

- better conform to common coding-styles: -pedantic -Wall -W -Wno-unused-parameter -std=gnu99

- check for duplicate ip configuration on local node

- remove lots of global variables

- plugin framework allowing registration of:
	- parameters
	- data-structures
	- scheduled functions
	- ...
	

- support muting of selected debug messages

- new bit-array processing funtions 

- remove usage of libpthread

- use ip address of primary interface as preferred src for all interfaces

- periodically check for changed ip_forward and other kernel configs


os.h
/*
 * Copyright (C) 2006 BATMAN contributors:
 * Thomas Lopatic, Marek Lindner, Axel Neumann
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of version 2 of the GNU General Public
 * License as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA
 *
 */

#ifndef _BATMAN_OS_H
#define _BATMAN_OS_H


/* get_time functions MUST be called at least every 2*MAX_SELECT_TIMEOUT_MS to allow for properly working time-drift checks */

/* overlaps after approximately 138 years */
//#define get_time_sec()  get_time( NO, NULL  ) 

/* overlaps after 49 days, 17 hours, 2 minutes, and 48 seconds */
//#define get_time_msec() get_time( YES, NULL ) 

# define timercpy(d, a) (d)->tv_sec = (a)->tv_sec; (d)->tv_usec = (a)->tv_usec; 


/* posix.c */

enum {
	CLEANUP_SUCCESS,
	CLEANUP_FAILURE,
	CLEANUP_MY_SIGSEV,
	CLEANUP_RETURN
};

void bat_wait( uint32_t sec, uint32_t msec );

#ifndef NOTRAILER
void print_animation( void );
#endif

int8_t send_udp_packet( unsigned char *packet_buff, int32_t packet_buff_len, struct sockaddr_in *dst, int32_t send_sock );

void cleanup_all( int status );

void update_batman_time( struct timeval *precise_tv );

char *get_human_uptime( uint32_t reference );

#ifndef NODEPRECATED
void fake_start_time( int32_t fake );
#endif


int32_t rand_num( uint32_t limit );


int8_t is_aborted();
//void handler( int32_t sig );
//void restore_and_exit( uint8_t is_sigsegv );

uint8_t get_set_bits( uint32_t v );




/* route.c */

#define DEV_LO "lo"
#define DEV_UNKNOWN "unknown"

#define	MIN_MASK	1
#define	MAX_MASK	32
#define ARG_MASK	"netmask"
#define ARG_NETW	"network"

extern int32_t base_port;
#define ARG_BASE_PORT "base_port"
#define DEF_BASE_PORT 4305
#define MIN_BASE_PORT 1025
#define MAX_BASE_PORT 60000




/***
 *
 * Things you should leave as is unless your know what you are doing !
 *
 * RT_TABLE_INTERFACES	routing table for announced (non-primary) interfaces IPs and other unique IP addresses
 * RT_TABLE_NETWORKS	routing table for announced networks
 * RT_TABLE_HOSTS	routing table for routes towards originators
 * RT_TABLE_TUNNEL	routing table for the tunnel towards the internet gateway
 * RT_PRIO_DEFAULT	standard priority for routing rules
 * RT_PRIO_UNREACH	standard priority for unreachable rules
 * RT_PRIO_TUNNEL	standard priority for tunnel routing rules
 *
 ***/



extern int32_t Rt_table;
#define ARG_RT_TABLE "rt_table_offset"
#define DEF_RT_TABLE 64
#define MIN_RT_TABLE 2
#define MAX_RT_TABLE 240

#define RT_TABLE_INTERFACES (Rt_table + 0)
#define RT_TABLE_NETWORKS   (Rt_table + 1)
#define RT_TABLE_HOSTS      (Rt_table + 2)
#define RT_TABLE_TUNNEL     (Rt_table + 4)


extern int32_t Rt_prio;
#define ARG_RT_PRIO "prio_rules_offset"
#define MIN_RT_PRIO 3
#define MAX_RT_PRIO 32765
#define DEF_RT_PRIO 6500

#define RT_PRIO_INTERFACES (Rt_prio + 0  )
#define RT_PRIO_HOSTS      (Rt_prio + 100)
#define RT_PRIO_NETWORKS   (Rt_prio + 199)
#define RT_PRIO_TUNNEL     (Rt_prio + 300)


extern int32_t prio_rules;
#define ARG_PRIO_RULES "prio_rules"


#define ARG_THROW_RULES "throw_rules"

#define ARG_NO_POLICY_RT "no_policy_routing"

#define ARG_PEDANTIC_CLEANUP "pedantic_cleanup"



extern uint8_t if_conf_soft_changed; // temporary enabled to trigger changed interface configuration
extern uint8_t if_conf_hard_changed; // temporary enabled to trigger changed interface configuration

extern int Mtu_min;


struct rules_node {
	struct list_head list;
	uint32_t network;
	uint8_t netmask;
	uint8_t rt_table;
	uint32_t prio;
	char *iif;
	int8_t rule_t;
	int8_t track_t;
};

 
struct routes_node {
	struct list_head list;
	uint32_t dest;
	uint8_t netmask;
	uint8_t rt_table;
	int8_t route_t;
	int8_t track_t;
};


/***
 *
 * route types: 0 = RTN_UNICAST, 1 = THROW, 2 = UNREACHABLE
 *
 ***/
#define RT_UNICAST	0
#define RT_THROW 	1
#define RT_UNREACH	2


//track types:
enum {
	TRACK_NO,
	TRACK_STANDARD,    //basic rules to interfaces, host, and networks routing tables
	TRACK_MY_HNA,
	TRACK_MY_NET,
	TRACK_OTHER_HOST,
	TRACK_OTHER_HNA, 
	TRACK_TUNNEL
};

void add_del_route( uint32_t dest, uint8_t netmask, uint32_t router, uint32_t source, int32_t ifi, char *dev, uint8_t rt_table, int8_t route_type, int8_t del, int8_t track );

/***
 *
 * rule types: 0 = RTA_SRC, 1 = RTA_DST, 2 = RTA_IIF
#define RTA_SRC 0
#define RTA_DST 1
#define RTA_IIF 2
 *
 ***/
 
void add_del_rule( uint32_t network, uint8_t netmask, uint8_t rt_table, uint32_t prio, char *iif, int8_t rule_type, int8_t del, int8_t track );

enum {
 IF_RULE_SET_TUNNEL,
 IF_RULE_CLR_TUNNEL,
 IF_RULE_SET_NETWORKS,
 IF_RULE_CLR_NETWORKS,
 IF_RULE_UPD_ALL,
 IF_RULE_CHK_IPS
};

int update_interface_rules( uint8_t cmd );


void check_kernel_config( struct batman_if *batman_if );

//int8_t bind_to_iface( int32_t sock, char *dev );

//int is_interface_up(char *dev);
void if_deactivate ( struct batman_if *batman_if );
void check_interfaces ();

void init_route( void );
void init_route_args( void );
void cleanup_route( void );


/* hna.c */

#define ARG_HNAS "hnas"


/* tunnel.c */

extern int32_t Gateway_class;
#define ARG_GWTUN_NETW "gateway_tunnel_network"
#define ARG_GATEWAYS "gateways"

#define ARG_RT_CLASS "routing_class"
#define ARG_GW_CLASS "gateway_class"

#ifndef	NOTUNNEL


#define ARG_UNRESP_GW_CHK "unresp_gateway_check"

#define ARG_TWO_WAY_TUNNEL "two_way_tunnel"

#define ARG_ONE_WAY_TUNNEL "one_way_tunnel"

#define ARG_GW_HYSTERESIS "gateway_hysteresis"
#define MIN_GW_HYSTERE    1
#define MAX_GW_HYSTERE    PROBE_RANGE/PROBE_TO100
#define DEF_GW_HYSTERE    2


#define BATMAN_TUN_PREFIX "bat"
#define MAX_BATMAN_TUN_INDEX 20 

#define TUNNEL_DATA 0x01
#define TUNNEL_IP_REQUEST 0x02
#define TUNNEL_IP_INVALID 0x03
#define TUNNEL_IP_REPLY 0x06

#define GW_STATE_UNKNOWN  0x01
#define GW_STATE_VERIFIED 0x02

#define ONE_MINUTE                60000

#define GW_STATE_UNKNOWN_TIMEOUT  (1  * ONE_MINUTE)
#define GW_STATE_VERIFIED_TIMEOUT (5  * ONE_MINUTE)

#define IP_LEASE_TIMEOUT          (1 * ONE_MINUTE)

#define MAX_TUNNEL_IP_REQUESTS 60 //12
#define TUNNEL_IP_REQUEST_TIMEOUT 1000 // msec

	
struct tun_request_type {
	uint32_t lease_ip;
	uint16_t lease_lt;
} __attribute__((packed));

struct tun_data_type {
	unsigned char ip_packet[MAX_MTU];
} __attribute__((packed));

struct tun_packet_start {
#if __BYTE_ORDER == __LITTLE_ENDIAN
	unsigned int type:4;
	unsigned int version:4;  // should be the first field in the packet in network byte order
#elif __BYTE_ORDER == __BIG_ENDIAN
	unsigned int version:4;
	unsigned int type:4;
#else
# error "Please fix <bits/endian.h>"
#endif
} __attribute__((packed));

struct tun_packet
{
	uint8_t  reserved1;
	uint8_t  reserved2;
	uint8_t  reserved3;

	struct tun_packet_start start;
#define TP_TYPE  start.type
#define TP_VERS  start.version	

	union
	{
		struct tun_request_type trt;
		struct tun_data_type tdt;
	}tt;
#define LEASE_IP  tt.trt.lease_ip
#define LEASE_LT  tt.trt.lease_lt
#define IP_PACKET tt.tdt.ip_packet
} __attribute__((packed));


#define TX_RP_SIZE (sizeof(struct tun_packet_start) + sizeof(struct tun_request_type))
#define TX_DP_SIZE (sizeof(struct tun_packet_start) + sizeof(struct tun_data_type))


struct gwc_args {
	uint32_t gw_state_stamp;
	uint8_t gw_state;
	uint8_t prev_gw_state;
	uint32_t orig;
	struct gw_node *gw_node;	// pointer to gw node
	struct sockaddr_in gw_addr;	// gateway ip
	char  gw_str[ADDR_STR_LEN];	// string of gateway ip
	struct sockaddr_in my_addr; 	// primary_ip
	uint32_t my_tun_addr;		// ip used for bat0 tunnel interface
	char  my_tun_str[ADDR_STR_LEN];	// string of my_tun_addr
	int32_t mtu_min;
	uint8_t tunnel_type;
	int32_t udp_sock;
	int32_t tun_fd;
	int32_t tun_ifi;
	char tun_dev[IFNAMSIZ];		// was tun_if
	uint32_t tun_ip_request_stamp;
	uint32_t tun_ip_lease_stamp;
	uint32_t tun_ip_lease_duration;
	uint32_t send_tun_ip_requests;
	uint32_t pref_addr;
	//	uint32_t last_invalidip_warning;
};


struct gws_args
{
	int8_t netmask;
	int32_t port;
	int32_t owt;
	int32_t twt;
	int32_t lease_time;
	int mtu_min;
	uint32_t my_tun_ip;
	uint32_t my_tun_netmask;
	uint32_t my_tun_ip_h;
	uint32_t my_tun_suffix_mask_h;
	struct sockaddr_in  client_addr;
	struct gw_client **gw_client_list;
	int32_t sock;
	int32_t tun_fd;
	int32_t tun_ifi;
	char tun_dev[IFNAMSIZ];
};


#define DEF_GWTUN_NETW_PREFIX  "169.254.0.0" /* 0x0000FEA9 */

#define MIN_GWTUN_NETW_MASK 20
#define MAX_GWTUN_NETW_MASK 30
#define DEF_GWTUN_NETW_MASK 22

#define MIN_TUN_LTIME 60 /*seconds*/
#define MAX_TUN_LTIME 60000
#define DEF_TUN_LTIME 600
#define ARG_TUN_LTIME "tunnel_lease_time"


#define MIN_RT_CLASS 0
#define MAX_RT_CLASS 3

// field accessor and flags for gateway announcement extension packets
#define EXT_GW_FIELD_GWTYPES ext_related
#define EXT_GW_FIELD_GWFLAGS def8
#define EXT_GW_FIELD_GWPORT  d16.def16
#define EXT_GW_FIELD_GWADDR  d32.def32

// the flags for gw extension messsage gwtypes:
#define TWO_WAY_TUNNEL_FLAG   0x01
#define ONE_WAY_TUNNEL_FLAG   0x02

struct tun_orig_data {
	
	int16_t  tun_array_len;
	struct ext_packet tun_array[];
	
};

struct plugin_v1 *tun_get_plugin_v1( void );
#endif

#endif
